### DISCUSSION TITLE
Better support for cancellation/interruption.

### DISCUSSION THREAD

**Discussion by ioquatix** [sidekiq/sidekiq#5127] [Ideas]
@fr2019 is working on adding support for job (and batch job) cancellation. We ended up implementing an entire UI to support it, but I pushed back on this gently suggesting that this is something that should be provided directly by Sidekiq & Sidekiq's UI/web interface. Otherwise we are going to end up implementing this same bespoke feature set across all applications where it's needed.

I would like to propose that we discuss and outline a way we can support this generally within Sidekiq. I think the scope encompasses:

- A mixin for cancellation using `Interrupt` exceptions or something similar.
- A mechanism for bulk cancellation of batch jobs.
- Extensions to the UI for handling cancellation and cancellation status.

I'm sure @fr2019 can fill in more details here, I'm just getting the conversation started.

**Comment by mperham** [comment]
Some questions I have:

1. Can a block of Ruby code be [safely "stopped" by an outside force](https://www.mikeperham.com/2015/05/08/timeout-rubys-most-dangerous-api/)? Java [documents its `Thread.stop` method as unsafe](https://docs.oracle.com/javase/1.5.0/docs/guide/misc/threadPrimitiveDeprecation.html). Sidekiq uses `Thread#kill` to halt long-running jobs which go over the shutdown timeout but only as a last resort because the process is exiting very soon. Should we be worried about transactions, connection pool leaks or any other side effects of non-local jumps?
2. Should user or job code be aware of and able to act on cancellation?
3. If you cancel a job, should that count as a success for Batch purposes?
4. Batches already have an API to invalidate the jobs in them, which prevents further execution but does not stop jobs in progress. How does that fit in here?

**Reply by ioquatix** [reply]
1. Use `thread.raise(Interrupt)` which is effectively the same as a user pressing Ctrl-C.
2. This will trigger ensure blocks and correctly terminate user code. However you need to be aware that isn't guaranteed to terminate use code, so it's a best effort (unlike `Thread#kill` which I don't think does any clean up at all and I'd personally advise you to avoid except as a very last resort).
3. My initial feeling is no, but you might want to track cancellations separately.
4. In the `ensure` block, you should propagate the cancellation IMHO.

**Comment by ciscoriordan** [comment]
The implementation I've been working on doesn't use batches. It allows cancelling an individual job, as shown in the FAQ (https://github.com/mperham/sidekiq/wiki/FAQ#how-do-i-cancel-a-sidekiq-job) with a similar `cancel_all!` method for cancelling all jobs for a given worker class. It's useful as a "big red button" in logjam situations.

These methods don't actually halt already running jobs, they just cause queued jobs to return early once run.

**Comment by mperham** [comment]
If you have a PR or something I can code review, that's one possible next step.

**Reply by ioquatix** [reply]
We don't have any PR, we built a bespoke interface internally.
